plugins/inputs/suricata/suricata.go

```diff
@@ -0,0 +1,224 @@
+// to Suricata.
+func (s *Suricata) Stop() {
+	s.inputListener.Close()
+	if s.cancel != nil {
+		s.cancel()
+	}
+	s.wg.Wait()
+	s.inputListener = nil
+}
+
+func (s *Suricata) readInput(ctx context.Context, acc telegraf.Accumulator, conn net.Conn) {
+	reader := bufio.NewReaderSize(conn, InBufSize)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			line, rerr := reader.ReadBytes('\n')
+			if rerr == io.EOF {
+				return
+			} else if len(line) > 0 {
+				s.parse(acc, line)
+			}
+		}
+	}
+}
+
+func (s *Suricata) handleServerConnection(ctx context.Context, acc telegraf.Accumulator) {
+	var err error
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			var conn net.Conn
+			conn, err = s.inputListener.Accept()
+			if err != nil {
+				if !strings.HasSuffix(err.Error(), ": use of closed network connection") {
+					acc.AddError(err)
+				}
+				return
+			}
+			s.readInput(ctx, acc, conn)
+		}
+	}
+}
+
+func flexFlatten(outmap map[string]interface{}, field string, v interface{}, delimiter string) error {
+	switch t := v.(type) {
+	case map[string]interface{}:
+		for k, v := range t {
+			var err error
+			if field == "" {
+				err = flexFlatten(outmap, k, v, delimiter)
+			} else {
+				err = flexFlatten(outmap, fmt.Sprintf("%s%s%s", field, delimiter, k), v, delimiter)
+			}
+			if err != nil {
+				return err
+			}
+		}
+	case float64:
+		outmap[field] = v.(float64)
+	default:
+		return fmt.Errorf("unsupported type %T encountered", t)
+	}
+	return nil
+}
+
+func (s *Suricata) parse(acc telegraf.Accumulator, sjson []byte) {
+	// initial parsing
+	var result map[string]interface{}
+	err := json.Unmarshal([]byte(sjson), &result)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	// check for presence of relevant stats
+	if _, ok := result["stats"]; !ok {
+		acc.AddError(fmt.Errorf("input does not contain necessary 'stats' sub-object"))
+		return
+	}
+
+	if _, ok := result["stats"].(map[string]interface{}); !ok {
+		acc.AddError(fmt.Errorf("'stats' sub-object does not have required structure"))
+		return
+	}
+
+	fields := make(map[string](map[string]interface{}))
+	totalmap := make(map[string]interface{})
+	for k, v := range result["stats"].(map[string]interface{}) {
+		if k == "threads" {
+			if v, ok := v.(map[string]interface{}); ok {
+				for k, t := range v {
+					outmap := make(map[string]interface{})
+					if threadStruct, ok := t.(map[string]interface{}); ok {
+						err = flexFlatten(outmap, "", threadStruct, s.Delimiter)
+						if err != nil {
+							acc.AddError(err)
+							// we skip this thread as something did not parse correctly
+							continue
+						}
+						fields[k] = outmap
+					}
+				}
+			} else {
+				acc.AddError(fmt.Errorf("'threads' sub-object does not have required structure"))
+			}
+		} else {
+			err = flexFlatten(totalmap, k, v, s.Delimiter)
+			if err != nil {
+				acc.AddError(err)
+				// we skip this subitem as something did not parse correctly
```

Comment:
True. There might be parts missing due to parse errors happening down the recursive calls. However, I wouldn't consider this a dealbreaker as:
- unlikely to happen anyway, Suricata produces well-formed JSON
- flattened totalmap is still valid, any error is tracked so there is a notion of something having failed

Reply:
You don't need to change this but I think it is simpler and more predicable if you return the error and stop processing.  It is substantially harder to do the right thing in the face of unexpected data and it might produce unreliable data.